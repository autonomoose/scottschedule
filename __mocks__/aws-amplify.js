export const { Hub } = jest.requireActual('aws-amplify');

const mockDefaultCurrentSession = {
    idToken: { payload: {'custom:tenant': 'test'} },
    accessToken: { payload: {
            username: 'mockuname',
            sub: 'mocksub'
}}};



export const Auth = {
    currentSession: jest.fn(() => Promise.resolve(mockDefaultCurrentSession)),
    currentAuthenticatedUser: jest.fn(() => Promise.resolve({
        username: 'testuser1',
        deleteUser: jest.fn(() => Promise.resolve('deleted')),
    })),
    userAttributes: jest.fn(() => Promise.resolve([
        {Name: 'sub', Value: 'test'},
        {Name: 'email_verified', Value: 'true'},
        {Name: 'email', Value: 'testuser1@test.com'},
        {Name: 'custom:tenant', Value: 'test'},
        {Name: 'phone_number_verified', Value: 'false'},
        {Name: 'phone_number', Value: '111'},
    ])),
    updateUserAttributes: jest.fn(() => Promise.resolve()),
    verifyCurrentUserAttribute: jest.fn(() => Promise.resolve()),
    verifyCurrentUserAttributeSubmit: jest.fn(() => Promise.resolve()),
    signIn: jest.fn(() => Promise.resolve()),
    signOut: jest.fn(() => Promise.resolve()),
    changePassword: jest.fn((currUsr, oldpass, newpass) => Promise.resolve('changed')),
};

export const Storage = {
    configure: jest.fn(() => Promise.resolve('origMockfile')),
    get: jest.fn(() => Promise.resolve('origMockfile')),
    put: jest.fn(() => Promise.resolve('origMockfile')),
    remove: jest.fn(() => Promise.resolve('removed origMockfile')),
    list: jest.fn(() => Promise.resolve([
        {key: "public/", size: 0},
        {key: "public/testlabel1", size: 23470},
        {key: "public/testlabel2", size: 22270},
    ]))
}

// used by layout before every authenticated page
export const mockAPIgqlGetCurrentUser = {
    'data': {
        'getCurrentUser': {
            adisable: null, agroups: null,
            nname: "user3test - legacy primary",
            perm: "standard", userid: "mockgraphql"
}}};
export const API = {
    post: jest.fn(() => Promise.resolve({'message': 'passed'})),
    graphql: jest.fn(() => Promise.resolve(mockAPIgqlGetCurrentUser)),
};

