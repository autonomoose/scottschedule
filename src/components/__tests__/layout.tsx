import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import { API, Auth, Hub } from 'aws-amplify';
import Layout from "../layout";

jest.mock('aws-amplify');

const mockEnqueue = jest.fn();
jest.mock('notistack', () => ({
  ...jest.requireActual('notistack'),
  useSnackbar: () => {
    return {
      enqueueSnackbar: mockEnqueue
    };
  }
}));

// mock the window.location.reload() called by auth HUB events
const { reload } = window.location;
beforeAll(() => {
    Object.defineProperty(window, 'location', {
        configurable: true,
        value: { reload: jest.fn() },
    });
});

afterAll(() => {
    window.location.reload = reload;
});

const mytest = <Layout><div data-testid='test'>Test</div></Layout>;
describe("Layout", () => {
  it("renders snapshot correctly", async () => {
    const {container, getByTestId} = render(mytest);
    await waitFor(() => {
        expect(getByTestId('test')).toBeVisible();
    });
    expect(container.firstChild).toMatchSnapshot();
  }, 10000);

  it('checks Error Boundary', async () => {
    const ThrowError = () => {
      throw new Error('Test error throw');
      return(<div>Mock Error</div>);
    };
    const consoleWarnFn = jest.spyOn(console, 'warn').mockImplementation(() => jest.fn());
    const consoleErrorFn = jest.spyOn(console, 'error').mockImplementation(() => jest.fn());

    render(<Layout><ThrowError /></Layout>);
    await waitFor(() => {
        expect(screen.getByTestId('errorboundary')).toBeVisible();
    });
    expect(consoleErrorFn).toHaveBeenCalledTimes(2); // error boundary and retry
    expect(consoleWarnFn).toHaveBeenCalledTimes(1);
    consoleErrorFn.mockRestore();
    consoleWarnFn.mockRestore();

  });

  it('checks auth Hub listener', async () => {
      const utils = render(mytest);
      await waitFor(() => {
          expect(utils.getByTestId('test')).toBeVisible();
      });
      const consoleLogFn = jest.spyOn(console, 'log').mockImplementation(() => jest.fn());
      const consoleWarnFn = jest.spyOn(console, 'warn').mockImplementation(() => jest.fn());

      Hub.dispatch('auth', {event: 'signOut', message: '', data:{}});
      await waitFor(() => {
          expect(mockEnqueue).toHaveBeenLastCalledWith(`user logged off`, {variant: 'success'});
      });

      Hub.dispatch('auth', {event: 'signIn', message: '', data:{}});
      await waitFor(() => {
          expect(mockEnqueue).toHaveBeenLastCalledWith(`Sign-on successful`, {variant: 'success'});
      });

      Hub.dispatch('auth', {event: 'tokenRefresh_failure', message: '', data:{}});
      await waitFor(() => {
          expect(mockEnqueue).toHaveBeenLastCalledWith(`user timed out`, {variant: 'success'});
      });

      Hub.dispatch('auth', {event: 'tokenRefresh', message: '', data:{}});
      expect(consoleLogFn).toHaveBeenLastCalledWith('user refreshed session');

      Hub.dispatch('auth', {event: 'mockUnknown', message: '', data:{}});
      expect(consoleLogFn).toHaveBeenLastCalledWith("Uncaught Auth module hub signal", "mockUnknown");

      // each dispatch event
      expect(consoleWarnFn).toHaveBeenCalledTimes(5);
      consoleWarnFn.mockRestore();
      consoleLogFn.mockRestore();
  });
  it('handles auth currentSession error throw', async () => {
      // patch in error
      const prevSession = Auth.currentSession;
      Auth.currentSession = jest.fn(() => Promise.reject('mockReject'));

      const utils = render(mytest);
      await waitFor(() => {
          expect(utils.getByTestId('authentFail')).toBeVisible();
      });

      Auth.currentSession = prevSession;
  });

  it('handles dberror during user lookup', async () => {
      const prevAPIgraphql = API.graphql;
      API.graphql = jest.fn(() => Promise.reject('mockRejected'));
      const utils = render(mytest);
      await waitFor(() => {
          expect(utils.getByTestId('authentFail')).toBeVisible();
      });
      API.graphql = prevAPIgraphql;

  });

  it('handles null lookup (new user)', async () => {
      const prevAPIgraphql = API.graphql;
      API.graphql = jest.fn(() => Promise.resolve({'data': {'getCurrentUser':{}}}));
      const utils = render(mytest);
      await waitFor(() => {
          expect(utils.getByTestId('authentNewUser')).toBeVisible();
      });
      API.graphql = prevAPIgraphql;

  });

});
