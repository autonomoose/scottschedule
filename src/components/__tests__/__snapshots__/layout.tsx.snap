// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout renders snapshot correctly 1`] = `
<div
  data-testid="authenticator"
>
   
  <div
    style="margin: 1rem auto; min-height: 100vh; text-align: center;"
  >
    <div
      class="MuiBox-root css-0"
    >
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorTransparent MuiAppBar-positionFixed mui-fixed css-1bzzxac-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1ti0wu3-MuiToolbar-root"
        >
          <button
            aria-label="Open menu"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeStart MuiIconButton-sizeMedium css-1t7w49f-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="MenuIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <a
            href="/home"
          >
            <div
              class="MuiBox-root css-k008qs"
            >
              <img
                alt="Werner Digital"
                src="../images/wernerdigital-hosted.png"
                width="26"
              />
              <p
                class="MuiTypography-root MuiTypography-body1 css-4peds9-MuiTypography-root"
              >
                Scottschedule
              </p>
            </div>
          </a>
          <div
            class="MuiBox-root css-1r3535z"
          >
            <a
              href="/usermaint"
              style="margin-right: 1em;"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                mockuname
              </p>
            </a>
            <button
              class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1vbvugf-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Sign Out
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <button
            aria-label="Dark/light mode"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeStart MuiIconButton-sizeMedium css-1t7w49f-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="LightModeIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zM2 13h2c.55 0 1-.45 1-1s-.45-1-1-1H2c-.55 0-1 .45-1 1s.45 1 1 1zm18 0h2c.55 0 1-.45 1-1s-.45-1-1-1h-2c-.55 0-1 .45-1 1s.45 1 1 1zM11 2v2c0 .55.45 1 1 1s1-.45 1-1V2c0-.55-.45-1-1-1s-1 .45-1 1zm0 18v2c0 .55.45 1 1 1s1-.45 1-1v-2c0-.55-.45-1-1-1s-1 .45-1 1zM5.99 4.58c-.39-.39-1.03-.39-1.41 0-.39.39-.39 1.03 0 1.41l1.06 1.06c.39.39 1.03.39 1.41 0s.39-1.03 0-1.41L5.99 4.58zm12.37 12.37c-.39-.39-1.03-.39-1.41 0-.39.39-.39 1.03 0 1.41l1.06 1.06c.39.39 1.03.39 1.41 0 .39-.39.39-1.03 0-1.41l-1.06-1.06zm1.06-10.96c.39-.39.39-1.03 0-1.41-.39-.39-1.03-.39-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06zM7.05 18.36c.39-.39.39-1.03 0-1.41-.39-.39-1.03-.39-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </header>
      <div
        class="MuiDrawer-root MuiDrawer-docked css-7ik032-MuiDrawer-docked"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft css-1hrb81m-MuiPaper-root-MuiDrawer-paper"
          style="visibility: hidden; webkit-transform: translateX(-0px); transform: translateX(-0px);"
        >
          <div
            class="MuiBox-root css-1ecnymn"
          >
            <div
              class="MuiBox-root css-xwg089"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
              >
                MainÂ Menu
              </h6>
              <button
                aria-label="Close menu"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1drgtl0-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ArrowDropUpIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m7 14 5-5 5 5z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-rr7dug-MuiDivider-root"
          />
          <ul
            class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
          >
            <a
              href="/home"
            >
              <div
                class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button css-ltlrkn-MuiButtonBase-root-MuiListItem-root"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemIcon-root css-qklaqr-MuiListItemIcon-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="HomeIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                  >
                    Home
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </div>
            </a>
            <a
              href="/usermaint"
            >
              <div
                class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button css-ltlrkn-MuiButtonBase-root-MuiListItem-root"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemIcon-root css-qklaqr-MuiListItemIcon-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="StarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                  >
                    Account
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </div>
            </a>
            <a
              href="/scheds"
            >
              <div
                class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button css-ltlrkn-MuiButtonBase-root-MuiListItem-root"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemIcon-root css-qklaqr-MuiListItemIcon-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="StarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                  >
                    Schedules
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </div>
            </a>
            <a
              href="/events"
            >
              <div
                class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button css-ltlrkn-MuiButtonBase-root-MuiListItem-root"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemIcon-root css-qklaqr-MuiListItemIcon-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="StarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                  >
                    Events
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </div>
            </a>
            <a
              href="/help"
            >
              <div
                class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button css-ltlrkn-MuiButtonBase-root-MuiListItem-root"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemIcon-root css-qklaqr-MuiListItemIcon-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="StarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                  >
                    Help
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </div>
            </a>
            <div
              class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button css-ltlrkn-MuiButtonBase-root-MuiListItem-root"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemIcon-root css-qklaqr-MuiListItemIcon-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="StarIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                  />
                </svg>
              </div>
              <div
                class="MuiListItemText-root css-tlelie-MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                >
                  Logout
                </span>
              </div>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </div>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-h8m1q6"
    >
      <main>
        <div
          data-testid="mainPageDisplay"
          style="text-align: left;"
        >
          <div
            data-testid="test"
          >
            Test
          </div>
        </div>
      </main>
    </div>
    <footer
      style="padding-top: 40px;"
    >
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-rr7dug-MuiDivider-root"
      />
      <div
        class="MuiBox-root css-1bnxa12"
      >
        <a
          href="/home"
        >
          Home
        </a>
        <a
          href="/usermaint"
        >
          Account
        </a>
        <a
          href="/scheds"
        >
          Schedules
        </a>
        <a
          href="/events"
        >
          Events
        </a>
        <a
          href="/help"
        >
          Help
        </a>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-rr7dug-MuiDivider-root"
      />
      <span
        class="MuiTypography-root MuiTypography-caption css-3merus-MuiTypography-root"
      >
        Â© 2021-
        2022
        , Werner Digital Technology Inc
      </span>
    </footer>
  </div>
</div>
`;
